File: api/src/main/java/nutri/cam/api/RecommendedPlate.java
package nutri.cam.api;

import java.util.HashSet;
import java.util.Set;
import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.Table;
import lombok.Data;

@Entity
@Table(name = "recommended_plate")
@Data
public class RecommendedPlate {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(nullable = false, unique = true, length = 150)
    private String name;

    @Column(nullable = false, length = 500)
    private String description;

    @Column(nullable = false, length = 200)
    private String portion;

    @Column(name = "estimated_calories")
    private Integer estimatedCalories;

    @Column(name = "estimated_protein")
    private Double estimatedProtein;

    @Column(name = "preparation_time_minutes")
    private Integer preparationTimeMinutes;

    @Enumerated(EnumType.STRING)
    @Column(name = "processing_level", nullable = false, length = 50)
    private ProcessingLevel processingLevel;

    @Column(nullable = false)
    private Boolean active = true;

    @Column(nullable = false, length = 100)
    private String source;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(
        name = "recommended_plate_meal_type",
        joinColumns = @JoinColumn(name = "recommended_plate_id")
    )
    @Enumerated(EnumType.STRING)
    @Column(name = "meal_type", nullable = false, length = 50)
    private Set<PlateMealType> allowedMealTypes = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(
        name = "recommended_plate_food_group",
        joinColumns = @JoinColumn(name = "recommended_plate_id")
    )
    @Enumerated(EnumType.STRING)
    @Column(name = "food_group", nullable = false, length = 50)
    private Set<FoodGroup> foodGroups = new HashSet<>();
}

File: nutricam_proyect/lib/models/recommended_plate.dart
class RecommendedPlate {
  final int id;
  final String name;
  final String description;
  final String portion;
  final int? estimatedCalories;
  final double? estimatedProtein;
  final int? preparationTimeMinutes;
  final String processingLevel;
  final bool active;
  final String source;
  final List<String> allowedMealTypes;
  final List<String> foodGroups;

  const RecommendedPlate({
    required this.id,
    required this.name,
    required this.description,
    required this.portion,
    required this.estimatedCalories,
    required this.estimatedProtein,
    required this.preparationTimeMinutes,
    required this.processingLevel,
    required this.active,
    required this.source,
    required this.allowedMealTypes,
    required this.foodGroups,
  });

  factory RecommendedPlate.fromJson(Map<String, dynamic> json) {
    return RecommendedPlate(
      id: json['id'] as int,
      name: json['name'] as String,
      description: json['description'] as String,
      portion: json['portion'] as String,
      estimatedCalories: json['estimatedCalories'] as int?,
      estimatedProtein: _parseDouble(json['estimatedProtein']),
      preparationTimeMinutes: json['preparationTimeMinutes'] as int?,
      processingLevel: json['processingLevel'] as String,
      active: json['active'] as bool,
      source: json['source'] as String,
      allowedMealTypes: _parseStringList(json['allowedMealTypes']),
      foodGroups: _parseStringList(json['foodGroups']),
    );
  }

  static double? _parseDouble(dynamic value) {
    if (value == null) {
      return null;
    }

    if (value is num) {
      return value.toDouble();
    }

    return double.tryParse(value.toString());
  }

  static List<String> _parseStringList(dynamic value) {
    if (value is! List) {
      return [];
    }

    return value.map((item) => item.toString()).toList();
  }
}